### PR TITLE
Adds match length to deterministic cache key

### LINKED ISSUE #533: Cache key to include Match length
Following #527 we have a nice cache class. This currently is designed for tournaments of constant match length and maps: pairs of players (when both players are not stochastic) to outcomes.

```
(axelrod.Cooperator, axelrod.Alternator): [('C', 'C'), ('C', 'D'), ('C', 'C')]
(axelrod.TitForTat, axelrod.Alternator): [('C', 'C'), ('C', 'D'), ('D', 'C')]
```

A small change would allow us to map pairs of players and match lengths to outcomes:

```
(axelrod.Cooperator, axelrod.Alternator, 3): [('C', 'C'), ('C', 'D'), ('C', 'C')]
(axelrod.TitForTat, axelrod.Alternator, 5): [('C', 'C'), ('C', 'D'), ('D', 'C'), ('C', 'D'), ('D', 'C')]
```

The main goal of this would be to continue the move away from assumptions about the tournament structure which will continue to make new tournaments easily. For example it would be easy to create a tournament where half matches are a single turn and the other half are a full 200 (or whatever).

Basically, whenever a match between deterministic players of any length happens, the cache could be used (or updated). The structure of the tournament would be irrelevant.

### PR DESCRIPTION
**Pull request by meatballs**
Closes #533

### CODE REVIEW

File: axelrod/deterministic_cache.py [hunk 1]
```diff
@@ -81,13 +77,18 @@ def _is_valid_key(self, key):
...
             return False
 
-        # The tuple should be a pair
-        if len(key) != 2:
+        # The tuple should be a triplet
+        if len(key) != 3:
             return False
 
-        # Each item in the pair should be a subclass of axelrod.Player
+        # The tripley should be a pair of axelrod.Player sublclasses and an
```
Comment (drvinceknight) [inline axelrod/deterministic_cache.py]: typo: "triplet"

File: axelrod/match.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ def __init__(self, players, turns, game=None, deterministic_cache=None,
         """
         self.result = []
         self.turns = turns
-        self._classes = (players[0].__class__, players[1].__class__)
+        self._cache_key = (players[0].__class__, players[1].__class__, turns)
```
Comment (drvinceknight) [inline axelrod/match.py]: Really like this rename.

**Comment by drvinceknight** [comment]
This looks awesome @meatballs. Will be able to use the cache for the prob end tournament now :)

Suggestion: instead of importing from future to deal with the py2 division could you just force float division in relevant places in the tests:

```
(3/float(turns), 2/float(turns))
```

Mainly because this is how it's done in a lot of other spots and also it's a bit more verbose?

**Comment by meatballs** [comment]
> Suggestion: instead of importing from future to deal with the py2 division could you just force float division in relevant places in the tests:

Yep, that's a better plan. I'll do that now

**Comment by drvinceknight** [comment]
Assuming travis is content, this is a big 👍 from me. Thanks @meatballs!

**Comment by meatballs** [comment]
> Assuming travis is content...

Always a big assumption when it comes to getting any my code working on python 2 and 3!

**Pull request merged**